### ISSUE TITLE
Dots in dependencies can cause confusing errors

### ISSUE THREAD

**Issue by viraptor** [python-poetry/poetry#1846]
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

If I try to add a dependency with a name like `zope.sqlalchemy` and make a mistake of not quoting the name, I get a confusing exception:

```
[KeyError]
'version'
```
(that's because the constraint is `{'sqlalchemy': '*'}`)

I know the definition is a mistake, but maybe a better handling of this exception could be possible - with a suggestion of what the problem is.

**Comment by Chais** [comment]
Still an issue, also for first-degree dependencies, where the error message is marginally more useful.  
To reproduce:
```
[tool.poetry]
name = "poetry_test"
version = "0.1.0"
description = ""
authors = []

[tool.poetry.dependencies]
python = "^2.7"
backports.functools-lru-cache = "^1.6.1"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```
Results in
```
$ poetry install

Python 2.7 will no longer be supported in the next feature release of Poetry (1.2).
You should consider updating your Python version to a supported one.

Note that you will still be able to manage Python 2.7 projects by using the env command.
See https://python-poetry.org/docs/managing-environments/ for more information.


RuntimeError

The Poetry configuration is invalid:
  - [dependencies.backports] {u'functools-lru-cache': u'^1.6.1'} is not valid under any of the given schemas
```

**Comment by sinoroc** [comment]
I think you need to use quotes:

```
[tool.poetry.dependencies]
python = "^2.7"
"backports.functools-lru-cache" = "^1.6.1"
```

**Comment by viraptor** [comment]
@sinoroc We know the fix (now, anyway). This issue is about how the error is presented, so that people don't get confused in the future.

**Comment by sinoroc** [comment]
> We know the fix (now, anyway). This issue is about how the error is presented, so that people don't get confused in the future.

Ah, right. My bad.

**Comment by sinoroc** [comment]
But **it is valid _TOML_**, right? So what should poetry do? I do not know if it is possible to catch these kinds of things, once the toml file is parsed. Maybe [_tomlkit_](https://github.com/sdispater/tomlkit) is able to figure that out somehow...

**Comment by finswimmer** [comment]
I agree with @sinoroc. Because it is a valid toml file, I have no clue how to catch this properly :thinking:

**Comment by sinoroc** [comment]
@finswimmer If I understood correcly _tomlkit_ has the relatively unique feature that it is aware of the formatting:

> It includes a parser that preserves all comments, indentations, whitespace and internal element ordering, and makes them accessible and editable via an intuitive API.

-- https://pypi.org/project/tomlkit/0.7.0/

So maybe it is possible to retrieve the information that the "_key_" `zope = { sqlalchemy = "*" }` in the  `dependencies` section (or `dev-dependencies`) is actually formatted as `zope.sqlalchemy = "*"`, so it would be possible to deduce that there is a user error with quoting.

Just brainstorming... I do not know, if it is possible or even worth the time and energy to look further into improving such a non-blocking thing.

**Comment by viraptor** [comment]
I would propose a bit of guesswork here. Basically if an entry in dependency section is a nested dictionary with the final value that's either `"*"` or looks like a version, then instead of "is not valid under any of the given schemas", say:

> this entry looks like a you missed quotes from the key, did you mean `"zope.sqlalchemy = ..."`

I'm not sure if a dictionary is ever allowed in that position though - if not, the check may be as simple as "is the value in dependencies section a dict?"

**Comment by Chais** [comment]
> I think you need to use quotes:
> 
> ```
> [tool.poetry.dependencies]
> python = "^2.7"
> "backports.functools-lru-cache" = "^1.6.1"
> ```

That may work for direct dependencies, but for higher degree dependencies it's not possible, I think, because the `poetry.lock` file is never created and thus can't be edited accordingly.

**Comment by viraptor** [comment]
@Chais If something actually fails to work with quoting, you should open a separate bug, otherwise it could get lost here. (this one's only about feedback format)
